internal/iso/iso.go

```diff
@@ -0,0 +1,188 @@
+	tinkWorkerImage = "tink_worker_image=quay.io/tinkerbell/tink-worker:v0.10.1"
+)
+
+type Handler struct {
+	Logger             logr.Logger
+	Backend            handler.BackendReader
+	SourceISO          string // assume to be an url for now
+	ExtraKernelParams  []string
+	Syslog             string
+	TinkServerTLS      bool
+	TinkServerGRPCAddr string
+	parsedURL          *url.URL
+	http.RoundTripper
+}
+
+type data struct {
+	MACAddress net.HardwareAddr
+	Facility   string
+}
+
+func (h *Handler) RoundTrip(req *http.Request) (*http.Response, error) {
+	h.Logger.Info("entered the roundtrip func")
+	notFoundResponse := &http.Response{
+		StatusCode: http.StatusNotFound,
+	}
+
+	if path.Base(req.URL.Path) != "hook.iso" {
+		h.Logger.Info("URL path not supported", "path", req.URL.Path)
+		return notFoundResponse, nil
+
+	}
+
+	ctx := req.Context()
+	// The incoming request url is expected to have the mac address present.
+	// Fetch the mac and validate if there's a hardware object
+	// associated with the mac.
+	//
+	// We serve the iso only if this validation passes.
+	ha, err := getMAC(req.URL.Path)
+	if err != nil {
+		h.Logger.Info("the hardware data for this machine is invalid", "client", ha, "error", err)
+		return notFoundResponse, nil
+
+	}
+
+	hw, err := getByMAC(ctx, ha, h.Backend)
+	if err != nil {
+		h.Logger.Info("retreiving the hardware object for this machine", "client", hw, "error", err)
+		return notFoundResponse, nil
+	}
+
+	// When the hook ISO is built, it is expected to have this random string embedded into its template
+	// in the kernel command line args
+	// the length of 268 was chosen to accommodate a variable number of args during runtime
+	//
+	// Ref: https://github.com/tinkerbell/hook/blob/main/linuxkit-templates/hook.template.yaml#L17
+	// TODO: For long term ease of use, we should incorporate this into a CLI flag
+	magicString := []byte(`#u8nda6ufbgp837zn07xb1wwh212gekw5bixoz5cy45y3krypw992nuofnocge0e345c2o98rqrp0g34n0qbum8n6ui12azvdve2mw5203j54jdr6i7uie8fzgd4bsexswah8a68fnabc6wvxo9xaau7k9kz1t28bc5rrg35xa48c1rji277jopd7v7xwd1sbz9xvzbdvyxtu9y6t2fom2s6ec3qvzkpz6h5y5yuti17ks74ugtsc3tdtfrzty37miqcsmeafim#`)
+	magicStringPadding := bytes.Repeat([]byte{' '}, len(magicString))
+	// The hardware object doesn't contain a specific field for consoles
+	// right now facility is used instead.
+	var consoles string
+	if hw.Facility != "" {
+		consoles = hw.Facility
+	} else {
+		consoles = defaultConsoles
+	}
+
+	// Reverse Proxy modifies the request url to
+	// the same path it received the incoming request.
+	// Manually override the path as the incoming request has mac address appended to the req path
+	// Ex: requesting <ip-address>:port/<mac-id>/hook.iso on the Source server would error out
+	req.URL.Path = h.parsedURL.Path
+	resp, err := h.RoundTripper.RoundTrip(req)
+	if err != nil {
+		h.Logger.Info("HTTP request didn't receive a response", "error", err)
```

Comment:
HTTP request to the `SourceISO` URL in this case. Might want to add the `SourceISO` to the keysAndValues and make the message part clear that this is the case too.